src/EventLoop.cpp

```
@@ -548,66 +512,75 @@ void EventLoop::Run(const std::chrono::milliseconds &duration)
 	while(running && now<=until)
 	{
 		//TRACE_EVENT("eventloop", "EventLoop::Run::Iteration");
-		ufds[0].events = GetPollEvents();
-		//Clear readed events
-		ufds[0].revents = 0;
-		ufds[1].revents = 0;
+		
+		poll->ForEachFd([this](int fd) {
+			
+			auto events = GetPollEventMask(fd);
+			if (events)
+				poll->SetEventMask(fd, *events);
+		});
 		
 		//Until signaled or one each 10 seconds to prevent deadlocks
 		int timeout = GetNextTimeout(10E3, until);
 
 		//UltraDebug(">EventLoop::Run() | poll timeout:%d timers:%d tasks:%d size:%d\n",timeout,timers.size(),tasks.size_approx(), sizeof(ufds) / sizeof(pollfd));
-		
+
 		//Wait for events
+		if (!poll->Wait(timeout))
 		{
-			//TRACE_EVENT("eventloop", "poll", "timeout", timeout);
-			(void)poll(ufds,sizeof(ufds)/sizeof(pollfd),timeout);
+			Error("Event poll wait failed.\n");
+			SetStopping(ToUType(PredefinedExitCode::WaitError));
+			break;
 		}
 		
 		//Update now
 		now = Now();
 		
 		//UltraDebug("<EventLoop::Run() | poll timeout:%d timers:%d tasks:%d\n",timeout,timers.size(),tasks.size_approx());
-
-		//Check for cancel
-		if ((ufds[0].revents & POLLHUP) || (ufds[0].revents & POLLERR) || (ufds[1].revents & POLLHUP) || (ufds[1].revents & POLLERR))
-		{
-			//Error
-			Log("-EventLoop::Run() | Pool error event [revents:%d,errno:%d]\n",ufds[0].revents,errno);
-			//Exit
-			break;
-		}
 		
-		//Read first
-		if (ufds[0].revents & POLLIN)
-		{
-			OnPollIn(ufds[0].fd);
-		}
+		auto [events, error] = poll->GetSignallingEvents();
```

could we return the error better directly on the `Wait()` ?

---

we don't need the events from the signaling pipe at all, right?

---

No. I can remove the events

---

I further removed the GetSignallingEvents(), and added the logic to the wait function. 
Also removed the ClearSignal function, now the clear is automatically done in the wait function.